src/mcts/search.cc

```diff
@@ -1466,7 +1441,88 @@ void SearchWorker::DoBackupUpdateSingleNode(
   search_->total_playouts_ += node_to_process.multivisit;
   search_->cum_depth_ += node_to_process.depth * node_to_process.multivisit;
   search_->max_depth_ = std::max(search_->max_depth_, node_to_process.depth);
-}  // namespace lczero
+}
+
+bool SearchWorker::MaybeConvertParent(Node* p, float m) const {
+  auto found_tb = false;
+  auto losing_m = 0.0f;
+
+  // Determine the maximum (Lower, Upper) bounds across all children.
+  // (-1,-1) Loss (initial and lowest bounds)
+  // (-1, 0) Can't Win
+  // (-1, 1) Regular node
+  // ( 0, 0) Draw
+  // ( 0, 1) Can't Lose
+  // ( 1, 1) Win (highest bounds)
+  auto max_L = -1;
+  auto max_U = -1;
+  for (const auto& edge : p->Edges()) {
+    if (edge.IsTerminal()) {
+      found_tb = found_tb || edge.IsTbTerminal();
+
+      // Default_wl doesn't matter as WL is only used if IsTerminal is true.
+      const auto WL = edge.GetWL(0.0f);
+      // We can short-circuit after a single win maxes the bounds.
+      if (WL > 0.0f) {
+        max_L = max_U = 1;
+        break;
+      } else if (WL == 0.0f) {
+        max_L = std::max(0, max_L);
+        max_U = std::max(0, max_U);
+      } else {
+        // Track the longest loss.
+        losing_m = std::max(losing_m, edge.GetM(0.0f));
+      }
+    } else if (edge.CantWin()) {
+      max_U = std::max(0, max_U);
+    } else if (edge.CantLose()) {
+      max_L = std::max(0, max_L);
+      max_U = 1;
+    } else {
+      // A regular or unvisited node maxes the upper bound.
+      max_U = 1;
+    }
+  }
+
+  auto parent_type =
+      found_tb ? Node::NodeType::TablebaseTerminal : Node::NodeType::Terminal;
```

Comment:
So even if we find tablebase loss and non-tablebase win, the node will still be marked as tablebase?.. I'm not sure that's intended.

Reply:
When finding a win, the first one causes the parent to become a loss and no more siblings will be visited. In the cases involving draws and losses, the best outcome is a draw which doesn't care about tablebase-ness. And for all losses, if any are tablebase, the converted parent win would assume the longest/tablebase loss is played.

But yes I see how the bug can arise now. A tablebase loss is found first (and ignored) and later iteration of search finds a low prior win. I guess we need to accumulate `found_win_tb` separately from `found_loss_tb` ?

Reply:
Added `found_mate` for preferring a checkmate (i.e., end of game win terminal) to then make the parent a terminal non-tb loss.